morpheus/pipeline/training/tao_client.py

```diff
@@ -0,0 +1,879 @@
+import os
+import tarfile
+import time
+import typing
+
+import requests
+
+logger = logging.getLogger(__name__)
+
+_KIND = typing.Literal["model", "dataset"]
+_DATASET_ACTIONS = typing.Literal["convert", "convert_index", "convert_efficientdet"]
+_MODEL_ACTIONS = typing.Literal["train", "evaluate", "prune", "retrain", "export", "inference"]
+
+
+def validate_kind(func):
+    """
+    Validates given endpoint category.
+
+    Parameters
+    ----------
+    func : Function that requires wrapping.
+    Returns
+    -------
+    inner_func
+        Encapsulated function.
+    """
+
+    @functools.wraps(func)
+    def inner_func(*args, **kwargs):
+        if len(args) < 2:
+            raise ValueError("Kind not found. Select from available kinds: {}".format(_KIND))
+        kind = args[1]
+
+        if kind is None:
+            raise TypeError("TypeError: a string-like object is required for kind, not 'NoneType'")
+        if kind not in typing.get_args(_KIND):
+            raise ValueError("Invalid kind '{}'. Available kinds are {}".format(kind, _KIND))
+        return func(*args, **kwargs)
+
+    return inner_func
+
+
+def validate_actions(func):
+    """
+    Validates TAO actions.
+
+    Parameters
+    ----------
+    func : Function that requires wrapping.
+    Returns
+    -------
+    inner_func
+        Encapsulated function.
+    """
+
+    @functools.wraps(func)
+    def inner_func(*args, **kwargs):
+
+        actions_by_kind = _DATASET_ACTIONS
+        if args[1] == "model":
+            actions_by_kind = _MODEL_ACTIONS
+
+        if len(args) < 3:
+            raise ValueError("Actions not found. Select from available actions: {}".format(actions_by_kind))
+
+        actions = args[2]
+
+        if actions is None:
+            raise TypeError("TypeError: a string-like object is required for an action, not 'NoneType'")
+
+        available_actions = typing.get_args(actions_by_kind)
+
+        if isinstance(actions, list):
+            if not set(actions).issubset(available_actions):
+                raise ValueError("One or more actions are not valid actions '{}'. Available actions are {}".format(
+                    actions, actions_by_kind))
+        else:
+            if actions not in available_actions:
+                raise ValueError("Invalid action '{}'. Available actions are {}".format(actions, actions_by_kind))
+
+        return func(*args, **kwargs)
+
+    return inner_func
+
+
+def generate_schema_url(url, ssl):
+    if url.startswith("http://") or url.startswith("https://"):
+        raise ValueError("URL should not include the scheme")
+
+    scheme = "https://" if ssl else "http://"
+    url = scheme + (url if url[-1] != "/" else url[:-1])
+
+    return url
+
+
+def vaildate_apikey(apikey):
+    if not isinstance(apikey, str):
+        raise ValueError('API key must be a string')
+
+    if not apikey:
+        raise ValueError('API key can not be an empty string')
+
+    return apikey
+
+
+class TaoApiClient():
```

Comment:
I dont think this is the correct location for this file. This class isnt related to pipelines at all and is standalone. Would work better in the `io` directory